stress/backend/src/main/scala/com/iheart/kanaloa/stress/backend/MockBackend.scala

```diff
@@ -13,69 +9,64 @@ import scala.util.Random
-        log.info("concurrent work: " + EntryActor.counter + ", new rate: " + EntryActor.counter)
-      } else if (EntryActor.counter >= maxRoutees) {
-        val rate = -math.pow(EntryActor.counter - maxRoutees, 2) / (0.5 * maxRoutees) + maxRoutees
-        val newRate = rate.toInt
-        if (newRate > 0) {
-          activeRoutees = throttlers.take(newRate)
-          log.info("concurrent work: " + EntryActor.counter + ", new rate: " + newRate)
-        } else {
-          activeRoutees = throttlers.take(1)
-          log.info("too much work, new rate: " + newRate)
-        }
+    val rand = new Random(System.currentTimeMillis())
+    val responders: Array[ActorRef] =
+      Array.tabulate(optimalConcurrency) { _ ⇒
+        val throttler = context.actorOf(Props(classOf[TimerBasedThrottler], (optimalThroughput.toDouble / 10d / optimalConcurrency).toInt msgsPer 100.milliseconds))
+        val responder = context.actorOf(Props(classOf[ResponderBehindThrottler]))
+        throttler ! SetTarget(Some(responder))
+        throttler
       }
-    }
+
+    def limitedRespondersWhenOverload =
+      (-math.pow(concurrentRequests - optimalConcurrency, 2) /
+        (0.5 * optimalConcurrency) + optimalConcurrency).toInt
 
     def receive = {
-      case Message(msg) ⇒
-        EntryActor.counter += 1
-        reThrottle()
-        log.info("message() received in Router")
-        val random_index = rand.nextInt(activeRoutees.length)
-        val responder = activeRoutees(random_index)
-        responder ! Petition(msg, sender, this.self)
+      case Request(msg) ⇒
+        concurrentRequests += 1
+
+        val activeResponders = if (concurrentRequests < optimalConcurrency)
+          concurrentRequests else limitedRespondersWhenOverload
+
+        log.info("concurrent work: " + concurrentRequests + ", responders: " + (activeResponders + 1))
+
+        val index =
+          if (activeResponders > 0)
+            rand.nextInt(activeResponders)
+          else 0
+
+        responders(index) ! Petition(msg, sender)
 
       case Appease(msg, replyTo) ⇒
-        EntryActor.counter -= 1
-        log.info("appease() received in Router")
-        replyTo ! Message(msg)
+        concurrentRequests -= 1
+        replyTo ! Respond(msg)
 
-      case _ ⇒ sender ! Message("don't recognize!")
+      case other ⇒ log.error("unknown msg received " + other)
     }
   }
 
-  class BackendResponder extends Actor with ActorLogging {
+  class ResponderBehindThrottler extends Actor with ActorLogging {
     import context.dispatcher
     def receive = {
-      case Petition(msg, roundTrip, source) ⇒
-        context.system.scheduler.scheduleOnce(50.milliseconds) {
-          log.info("petition() received in Responder")
-          source ! Appease(msg, roundTrip)
-        }
-      case _ ⇒ log.info("received unknown message in Responder")
+      case Petition(msg, replyTo) ⇒
+        context.system.scheduler.scheduleOnce(50.milliseconds, sender(), Appease(msg, replyTo))
```

Comment:
before you were probably using the sender in the `scheduleOnce` block, which will give you the error, you saw. because sender is actually a method, and by the time the `scheduleOnce` block runs it might change. So when you call send() you want to make sure it's called immediately in the receive block, not in an async block such as scheduleOnce or `Future.map`